File: api/ruby-sinatra-api/services/todo_service.rb
require_relative '../models/todo'

class TodoService
  def initialize
    @todos = {}
    load_sample_data
  end

  def get_all_todos(filter: nil, search: nil, priority: nil)
    todos = @todos.values

    # Apply filters
    todos = todos.select(&:completed) if filter == 'completed'
    todos = todos.reject(&:completed) if filter == 'active'
    todos = todos.select { |t| t.priority == priority } if priority
    todos = todos.select { |t| t.text.downcase.include?(search.downcase) } if search

    todos
  end

  def get_todo_by_id(id)
    @todos[id]
  end

  def create_todo(text:, priority: nil, completed: nil, due_date: nil, reminder_time: nil)
    todo = Todo.new(
      text: text,
      priority: priority || 'medium',
      completed: completed || false,
      due_date: due_date,
      reminder_time: reminder_time
    )
    @todos[todo.id] = todo
    todo
  end

  def update_todo(id, text: nil, priority: nil, completed: nil, due_date: nil, reminder_time: nil)
    todo = @todos[id]
    return nil unless todo

    todo.text = text if text
    todo.priority = priority if priority && Todo::VALID_PRIORITIES.include?(priority)
    todo.completed = completed unless completed.nil?
    todo.due_date = due_date unless due_date.nil?
    todo.reminder_time = reminder_time unless reminder_time.nil?
    todo.update_timestamp

    todo
  end

  def delete_todo(id)
    @todos.delete(id) ? true : false
  end

  def toggle_todo(id)
    todo = @todos[id]
    return nil unless todo

    todo.completed = !todo.completed
    todo.update_timestamp
    todo
  end

  def get_stats
    todos = @todos.values
    total = todos.size
    completed = todos.count(&:completed)
    active = total - completed

    priority_breakdown = {
      'low' => todos.count { |t| t.priority == 'low' },
      'medium' => todos.count { |t| t.priority == 'medium' },
      'high' => todos.count { |t| t.priority == 'high' }
    }

    overdue_count = todos.count(&:overdue?)
    due_today_count = todos.count(&:due_today?)
    upcoming_count = todos.count(&:due_soon?)

    {
      total: total,
      active: active,
      completed: completed,
      completionRate: total > 0 ? (completed.to_f / total * 100).round(2) : 0,
      priorityBreakdown: priority_breakdown,
      overdueCount: overdue_count,
      dueTodayCount: due_today_count,
      upcomingCount: upcoming_count
    }
  end

  def clear_completed
    @todos.reject! { |_, todo| todo.completed }
  end

  private

  def load_sample_data
    Todo.create_sample_todos.each do |todo|
      @todos[todo.id] = todo
    end
  end
end

File: api/dotnet-api/Services/TodoService.cs
using System.Collections.Concurrent;
using DotnetApi.Models;

namespace DotnetApi.Services
{
    public class TodoService
    {
        private readonly ConcurrentDictionary<string, Todo> _todos = new();

        public TodoService()
        {
            LoadSampleData();
        }

        public IEnumerable<Todo> GetAllTodos(string? filter = null, string? search = null, string? priority = null)
        {
            return _todos.Values.Where(todo => ApplyFilters(todo, filter, search, priority));
        }

        public Todo? GetTodoById(string id)
        {
            _todos.TryGetValue(id, out var todo);
            return todo;
        }

        public Todo CreateTodo(string text, Priority? priority = null, bool? completed = null, string? dueDate = null, string? reminderTime = null)
        {
            var todo = new Todo
            {
                Text = text,
                Priority = priority ?? Priority.medium,
                Completed = completed ?? false,
                DueDate = dueDate,
                ReminderTime = reminderTime
            };

            _todos[todo.Id] = todo;
            return todo;
        }

        public Todo? UpdateTodo(string id, string? text = null, Priority? priority = null, bool? completed = null, string? dueDate = null, string? reminderTime = null)
        {
            if (!_todos.TryGetValue(id, out var todo))
                return null;

            if (text != null) todo.Text = text;
            if (priority.HasValue) todo.Priority = priority.Value;
            if (completed.HasValue) todo.Completed = completed.Value;
            if (dueDate != null) todo.DueDate = dueDate;
            if (reminderTime != null) todo.ReminderTime = reminderTime;

            todo.UpdateTimestamp();
            return todo;
        }

        public bool DeleteTodo(string id)
        {
            return _todos.TryRemove(id, out _);
        }

        public Todo? ToggleTodo(string id)
        {
            if (!_todos.TryGetValue(id, out var todo))
                return null;

            todo.Completed = !todo.Completed;
            todo.UpdateTimestamp();
            return todo;
        }

        public TodoStats GetStats()
        {
            var todos = _todos.Values.ToList();
            var total = todos.Count;
            var completed = todos.Count(t => t.Completed);
            var active = total - completed;

            var priorityBreakdown = new Dictionary<string, int>
            {
                { "low", todos.Count(t => t.Priority == Priority.low) },
                { "medium", todos.Count(t => t.Priority == Priority.medium) },
                { "high", todos.Count(t => t.Priority == Priority.high) }
            };

            var today = DateTime.Today;
            var overdueCount = 0;
            var dueTodayCount = 0;
            var upcomingCount = 0;

            foreach (var todo in todos.Where(t => !t.Completed && !string.IsNullOrEmpty(t.DueDate)))
            {
                if (DateTime.TryParse(todo.DueDate, out var dueDate))
                {
                    if (dueDate.Date < today)
                        overdueCount++;
                    else if (dueDate.Date == today)
                        dueTodayCount++;
                    else if (dueDate.Date <= today.AddDays(7))
                        upcomingCount++;
                }
            }

            return new TodoStats
            {
                Total = total,
                Active = active,
                Completed = completed,
                CompletionRate = total > 0 ? (double)completed / total * 100 : 0,
                PriorityBreakdown = priorityBreakdown,
                OverdueCount = overdueCount,
                DueTodayCount = dueTodayCount,
                UpcomingCount = upcomingCount
            };
        }

        public void ClearCompleted()
        {
            var completedIds = _todos.Where(kvp => kvp.Value.Completed).Select(kvp => kvp.Key).ToList();
            foreach (var id in completedIds)
            {
                _todos.TryRemove(id, out _);
            }
        }

        private bool ApplyFilters(Todo todo, string? filter, string? search, string? priority)
        {
            if (filter == "active" && todo.Completed) return false;
            if (filter == "completed" && !todo.Completed) return false;
            if (!string.IsNullOrEmpty(priority) && todo.Priority.ToString() != priority) return false;
            if (!string.IsNullOrEmpty(search) && !todo.Text.Contains(search, StringComparison.OrdinalIgnoreCase)) return false;
            return true;
        }

        private void LoadSampleData()
        {
            var today = DateTime.Today;
            var tomorrow = today.AddDays(1).ToString("yyyy-MM-dd");
            var nextWeek = today.AddDays(7).ToString("yyyy-MM-dd");
            var yesterday = today.AddDays(-1).ToString("yyyy-MM-dd");

            CreateSampleTodo("Learn .NET Core and C#", Priority.high, false, tomorrow, "09:00");
            CreateSampleTodo("Build REST API with ASP.NET", Priority.high, true, yesterday, "14:30");
            CreateSampleTodo("Add Entity Framework integration", Priority.medium, false, nextWeek, "16:00");
            CreateSampleTodo("Write unit tests with xUnit", Priority.medium, false, null, null);
            CreateSampleTodo("Deploy to Azure", Priority.low, false, null, null);
        }

        private void CreateSampleTodo(string text, Priority priority, bool completed, string? dueDate, string? reminderTime)
        {
            var todo = new Todo
            {
                Text = text,
                Priority = priority,
                Completed = completed,
                DueDate = dueDate,
                ReminderTime = reminderTime
            };
            _todos[todo.Id] = todo;
        }
    }
}

